### PR TITLE
Feature #2540 Point2Grid config file

### LINKED ISSUE #2540: Enhancement: Support for setting Point2Grid MET config variables
## Describe the Enhancement ##
In this discussion, [Need Point2Grid metplus wrapper for EVSv2.0 #2539](https://github.com/dtcenter/METplus/discussions/2539),   [Ho-ChunHuang-NOAA](https://github.com/Ho-ChunHuang-NOAA) mentioned the need to use Point2Grid app to map GOES ABI AOD to model grid for EVS2 AOD verification for AQM, RAP, HRRR, and GEFS-Chem.   A Point2GridConfig_wrapped file does not currently exist at https://github.com/dtcenter/METplus/tree/develop/parm/met_config.  It seems that METplus does not currently include logic for populating and passing a config file to the MET point2grid tool.

Please note that currently, the [METplus documentation](https://metplus.readthedocs.io/en/develop/generated/met_tool_wrapper/Point2Grid/Point2Grid.html#met-configuration) states with regard to the MET Configuration, "None. Point2Grid does not use configuration files.", but the MET Point2Grid [usage statement](https://met.readthedocs.io/en/latest/Users_Guide/reformat_point.html#point2grid-usage) indicates that it supports an optional configuration file.  The METplus documentation is incorrect and should be updated.

### Time Estimate ###
~1 day

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
6.0.0-beta6

### Funding Source ###
TBD

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED CYCLE ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by georgemccabe**
Note: dtcenter/MET#2880 must be completed before this can be properly tested

## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

Added unit tests to ensure correct behavior of new config variables

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Review code changes and documentation changes. Ensure new METplus config variables correctly correspond to the appropriate MET settings. Optionally run an example using this branch with the latest version of MET.

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [X] Do these changes introduce new SonarQube findings? **[No]**</br>
If **yes**, please describe:

- [X] Please complete this pull request review by **10/8/2024**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [X] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/appendixA.html) table.
- [X] Review the source issue metadata (required labels, projects, and milestone).
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [X] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [X] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by georgemccabe** [comment]
After PR #2587 is merged, merge develop and add changes to support/document `time_offset_warning` for Point2Grid wrapper.

File: docs/Users_Guide/glossary.rst
Comment (georgemccabe) [inline docs/Users_Guide/glossary.rst:11633]:
```suggestion
     Specify the value for 'init_time' in the MET configuration file for Point2Grid.
```
You put the wrong name, silly.

Comment (georgemccabe) [reply docs/Users_Guide/glossary.rst:11633]: No it's not, it should be valid time

File: docs/Users_Guide/glossary.rst
Comment (JohnHalleyGotway) [inline docs/Users_Guide/glossary.rst]: @georgemccabe the `time_offset_warning` config option only applies to MET tools that do a grid-to-grid comparison. Doing some grepping, I only see calls to  this [time_offset_warning(...)](https://github.com/dtcenter/MET/blob/7d44b8db48dc79a47059252e50291c21000e1de0/src/basic/vx_config/config_file.cc#L329) function from wavelet_stat, mode, series_analysis, pcp_combine, grid_stat, and wwmca.

So it definitely doesn't impact point2grid. Not sure how to handle this.

File: docs/Users_Guide/wrappers.rst
Comment (JohnHalleyGotway) [inline docs/Users_Guide/wrappers.rst]: Not sure if we should add `POINT2GRID_TIME_OFFSET_WARNING` since it isn't actually being used.

**Review by JohnHalleyGotway — COMMENTED** [review]
I did leave a comment about the limited scope of the `time_offset_warning` message in MET compared to the broad implementation of it in METplus. So give that some thought.

I did review the code changes and don't see any issues. I do note that all GHA tests.

I also tested this on seneca:
```
cd /d1/projects/METplus/pull_requests/METplus-6.0.0/pr_2580/METplus
METplus/ush/run_metplus.py -c METplus/parm/use_cases/met_tool_wrapper/Point2Grid/Point2Grid.conf -c seneca.conf
METplus/ush/run_metplus.py -c METplus/parm/use_cases/met_tool_wrapper/Point2Grid/Point2Grid.conf -c seneca.conf config.POINT2GRID_GOES_QC_FLAGS='0'
```
However using `config.POINT2GRID_GOES_QC_FLAGS='0,1,2'` results in a parsing error:
```
ERROR: Invalid argument: 1
ERROR: Invalid argument: 2
```
But that's how it's run in this [MET unit test](https://github.com/dtcenter/MET/blob/7d44b8db48dc79a47059252e50291c21000e1de0/internal/test_unit/xml/unit_point2grid.xml#L205). Is there a way to specify a comma-separated list on the command line or does it need to be done in the config file?

**Comment by georgemccabe** [comment]
@JohnHalleyGotway, #511 is needed to support lists as command line overrides.

**Review by JohnHalleyGotway — DISMISSED** [review]
I approve of these changes. I tested this branch on seneca and the changes are working as expected.

I do recommend revisiting where the `..._TIME_OFFSET_WARNING` config options are supported. But I'm torn. It is kind of nice for them to be "global" so that we don't need to keep them in as lock-step between MET and METplus, but in reality, it only actually impacts a fraction of the MET tools. Not sure how to proceed.

**Comment by georgemccabe** [comment]
@JohnHalleyGotway, could you please review commit 9a2afbad to ensure that I removed time_offset_warning from the correct wrappers?

**Review by JohnHalleyGotway — DISMISSED** [review]
@georgemccabe these changes look good to me.

I reviewed the diffs in [this commit](https://github.com/dtcenter/METplus/commit/9a2afbadfa8c105ed7769338e71f4046fb5a0971) and see that `time_offset_warning` is *removed* from the wrappers for:
- madis2nc, ascii2nc, ensemble_stat, gen_ens_prod, grid_diag, ioda2nc, mtd, pb2nc, plot_point_obs, point_stat, stat_analysis, tc_diag, tc_gen, tc_pairs, tc_rmw, tc_stat, point2grid

And that `time_offset_warning` *remains* in the wrappers for:
- grid_stat, mode, series_analysis, wavelet_stat

Those changes are consistent with where `time_offset_warning` is actually referenced in the MET codebase (mode, series_analysis, wavelet_stat, grid_stat):
```
grep time_offset_warning `find ./ -name "*.cc"`
./src/basic/vx_config/config_file.cc:bool MetConfig::time_offset_warning(int offset)
./src/basic/vx_config/config_file.cc:int allowable_offset = lookup_int(conf_key_time_offset_warning, false);
./src/tools/core/mode/mode_exec.cc:      if(engine.conf_info.conf.time_offset_warning(
./src/tools/core/mode/mode_exec.cc:      if(engine.conf_info.conf.time_offset_warning(
./src/tools/core/mode/mode_exec.cc:      if(engine.conf_info.conf.time_offset_warning(
./src/tools/core/series_analysis/series_analysis.cc:      if(conf_info.conf.time_offset_warning(
./src/tools/core/wavelet_stat/wavelet_stat.cc:         if(conf_info.conf.time_offset_warning(
./src/tools/core/grid_stat/grid_stat.cc:         if(conf_info.conf.time_offset_warning(
./src/tools/core/pcp_combine/pcp_combine.cc:      if(config.time_offset_warning(
./src/tools/other/wwmca_tool/nc_output.cc:         if(Config->time_offset_warning(
./src/tools/other/wwmca_tool/nc_output.cc:         if(Config->time_offset_warning(
```

I note that the `pcp_combine` reference applies to the `-subtract` command and can be specified by the user it their config string.
I also note that no wwmca wrapper exists, so no updates are needed for that.

**Comment by georgemccabe** [comment]
I pushed a fix to allow the use cases to run successfully. @JohnHalleyGotway already approved the changes, so I am going to merge this without a 3rd approval.

**Pull request merged**
